### PR TITLE
Check for GUI eventloop when testing for jupyter kernel

### LINKED ISSUE #1708: Vispy wrongly detects Spyder as Jupyter notebook
https://github.com/vispy/vispy/blob/a6ec64bb7b56739b463e3c16245a31940ca4f5f7/vispy/app/application.py#L131

```python
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
Type "copyright", "credits" or "license" for more information.

IPython 7.8.0 -- An enhanced Interactive Python.

ip = get_ipython()

ip.has_trait('kernel')
Out[2]: True

ip
Out[3]: <ipykernel.zmqshell.ZMQInteractiveShell at 0x7f8c103ec2b0>

ip.config
Out[4]: 
{'InteractiveShell': {'xmode': 'Plain'},
 'IPCompleter': {'use_jedi': False, 'greedy': False},
 'IPKernelApp': {'exec_lines': ['',
   "import sys;sys.argv = [''];del sys",
   'get_ipython().kernel._load_autoreload_magic()',
   'get_ipython().kernel._load_wurlitzer()',
   "get_ipython().kernel._set_mpl_backend('qt5', False)"],
  'connection_file': '/run/user/1000/jupyter/kernel-7c84aada0ff3.json'},
 'InlineBackend': {'rc': {'figure.figsize': (6.0, 4.0),
   'figure.dpi': 72,
   'font.size': 10,
   'figure.subplot.bottom': 0.125,
   'figure.facecolor': 'white',
   'figure.edgecolor': 'white'}},
 'ZMQInteractiveShell': {'autocall': 0, 'banner1': ''}}
```

image

```python
vispy.__version__
Out[6]: '0.6.1'
```

@ccordoba12 any ideas?

@jni because I think you are using vispy these days????

### PR DESCRIPTION
**Pull request by hmaarrfk**
Fixes #1708

Do we need formal tests??

### CODE REVIEW

**Comment by djhoese** [comment]
If you have an idea for how to test this (we do have a bare minimum of tests for the notebook backend) then it would be appreciated. Having tests for this would also be great if/when jupyter changes how kernels/loops are stored or structured our tests should start failing.

**Comment by djhoese** [comment]
You might be able to write a simple test that creates an Application so that it uses the `_use` method and therefore call the notebook method; mocking what you have to.

**Comment by hmaarrfk** [comment]
@djhoese I don't think your old test is valid for jupyter console anymore
nor is it valid for jupyter qtconsole :/

```
[JupyterQtConsoleApp] WARNING | No such comm: c642ac22ffff4d6983713429dc80270e
(dev) mark@mark-xps:~/git/vispy$ jupyter console
Jupyter console 6.0.0

Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.



In [1]: get_ipython().active_eventloop                                          

In [2]: get_ipython().has_trait('kernel')                                       
Out[2]: True

In [3]: get_ipython().active_eventloop is None                                  
Out[3]: True

In [4]:   
```

<details>

```
(mcam_dev) mark@mark-xps:~/git/vispy$ conda list
# packages in environment at /home/mark/miniconda3/envs/mcam_dev:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
alabaster                 0.7.12                     py_0    conda-forge
apipkg                    1.5                        py_0    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
asn1crypto                1.2.0                    py37_0    conda-forge
astroid                   2.3.2                    py37_0    conda-forge
asv                       0.4.1            py37he1b5a44_0    conda-forge
atomicwrites              1.3.0                      py_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
autopep8                  1.4.4                      py_0    conda-forge
babel                     2.7.0                      py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
bleach                    3.1.0                      py_0    conda-forge
bokeh                     1.3.4                    py37_0    conda-forge
breathe                   4.13.0                     py_0    conda-forge
bzip2                     1.0.8                h516909a_1    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
cairo                     1.16.0            hfb77d84_1002    conda-forge
certifi                   2019.9.11                py37_0    conda-forge
cffi                      1.13.0           py37h8022711_0    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
clang                     9.0.0                hc9558a2_1    conda-forge
clang-tools               9.0.0                hc9558a2_1    conda-forge
clangdev                  9.0.0                hc9558a2_1    conda-forge
clangxx                   9.0.0                         1    conda-forge
click                     7.0                        py_0    conda-forge
cloudpickle               1.2.2                      py_0    conda-forge
cryptography              2.7              py37h72c5cf5_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
cython                    0.29.13          py37he1b5a44_0    conda-forge
cytoolz                   0.10.0           py37h516909a_0    conda-forge
dask                      2.6.0                      py_0    conda-forge
dask-core                 2.6.0                      py_0    conda-forge
dbus                      1.13.6               he372182_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
distributed               2.6.0                      py_0    conda-forge
docutils                  0.15.2                   py37_0    conda-forge
doxygen                   1.8.16               he8d5c6f_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
execnet                   1.7.1                      py_0    conda-forge
expat                     2.2.5             he1b5a44_1004    conda-forge
ffmpeg                    4.1.3                h167e202_0    conda-forge
filelock                  3.0.12                   pypi_0    pypi
flake8                    3.7.8                    py37_1    conda-forge
flit                      1.3                      py37_0    conda-forge
fontconfig                2.13.1            h86ecdb6_1001    conda-forge
freetype                  2.10.0               he983fc9_1    conda-forge
freetype-py               2.1.0.post1                py_0    conda-forge
fribidi                   1.0.5             h516909a_1002    conda-forge
fsspec                    0.5.2                      py_0    conda-forge
future                    0.18.1                   py37_0    conda-forge
gdown                     3.8.2                    pypi_0    pypi
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
giflib                    5.1.9                h516909a_0    conda-forge
glib                      2.58.3            h6f030ca_1002    conda-forge
gmp                       6.1.2             hf484d3e_1000    conda-forge
gnutls                    3.6.5             hd3a4fd2_1002    conda-forge
graphite2                 1.3.13            hf484d3e_1000    conda-forge
graphviz                  2.40.1               h0f2764d_1    conda-forge
gst-plugins-base          1.14.5               h0935bb2_0    conda-forge
gstreamer                 1.14.5               h36ae1b5_0    conda-forge
harfbuzz                  2.4.0                h9f30f68_3    conda-forge
hdf5                      1.10.5          nompi_h3c11f04_1103    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       64.2                 he1b5a44_1    conda-forge
idna                      2.8                   py37_1000    conda-forge
imagesize                 1.1.0                      py_0    conda-forge
importlib_metadata        0.23                     py37_0    conda-forge
ipdb                      0.12.2                     py_0    conda-forge
ipykernel                 5.1.2            py37h5ca1d4c_0    conda-forge
ipython                   7.8.0            py37h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
isort                     4.3.21                   py37_0    conda-forge
jasper                    1.900.1           h07fcdf6_1006    conda-forge
jedi                      0.15.1                   py37_0    conda-forge
jeepney                   0.4.1                      py_0    conda-forge
jinja2                    2.10.3                     py_0    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
json5                     0.8.5                      py_0    conda-forge
json_tricks               3.13.1                     py_0    conda-forge
jsonschema                3.1.1                    py37_0    conda-forge
jupyter                   1.0.0                      py_2    conda-forge
jupyter_client            5.3.3                    py37_1    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_core              4.5.0                      py_0    conda-forge
jupyterlab                1.1.4                      py_0    conda-forge
jupyterlab_server         1.0.6                      py_0    conda-forge
keyring                   19.2.0                   py37_0    conda-forge
kiwisolver                1.1.0            py37hc9558a2_0    conda-forge
lame                      3.100             h14c3975_1001    conda-forge
lazy-object-proxy         1.4.2            py37h516909a_0    conda-forge
libblas                   3.8.0               14_openblas    conda-forge
libcblas                  3.8.0               14_openblas    conda-forge
libclang                  9.0.0                hc9558a2_1    conda-forge
libffi                    3.2.1             he1b5a44_1006    conda-forge
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_2    conda-forge
libiconv                  1.15              h516909a_1005    conda-forge
liblapack                 3.8.0               14_openblas    conda-forge
liblapacke                3.8.0               14_openblas    conda-forge
libllvm9                  9.0.0                hc9558a2_2    conda-forge
libopenblas               0.3.7                h6e990d7_2    conda-forge
libpng                    1.6.37               hed695b0_0    conda-forge
libsodium                 1.0.17               h516909a_0    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.0.10            hfc65ed5_1004    conda-forge
libtool                   2.4.6             h14c3975_1002    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libwebp                   1.0.2                h576950b_1    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.9                hee79883_5    conda-forge
libxslt                   1.1.33               h31b3aaa_0    conda-forge
llvm-tools                9.0.0                hc9558a2_2    conda-forge
llvmdev                   9.0.0                hc9558a2_2    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.8.3             he1b5a44_1001    conda-forge
markupsafe                1.1.1            py37h14c3975_0    conda-forge
matplotlib-base           3.1.1            py37he7580a8_1    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
mistune                   0.8.4           py37h14c3975_1000    conda-forge
more-itertools            7.2.0                      py_0    conda-forge
msgpack-python            0.6.2            py37hc9558a2_0    conda-forge
mypy                      0.740                      py_0    conda-forge
mypy_extensions           0.4.3                    py37_0    conda-forge
nbconvert                 5.6.0                    py37_1    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
nettle                    3.4.1             h1bed415_1002    conda-forge
networkx                  2.4                        py_0    conda-forge
nose                      1.3.7                 py37_1002    conda-forge
notebook                  6.0.1                    py37_0    conda-forge
numpy                     1.17.2           py37h95a1406_0    conda-forge
numpydoc                  0.9.1                      py_0    conda-forge
olefile                   0.46                       py_0    conda-forge
opencv                    4.1.1            py37ha799480_1    conda-forge
openh264                  1.8.0             hdbcaa40_1000    conda-forge
openssl                   1.1.1c               h516909a_0    conda-forge
owl                       0.9.23+2.gb37242d.dirty           dev_0    <develop>
packaging                 19.2                       py_0    conda-forge
pandas                    0.25.2           py37hb3f55d8_0    conda-forge
pandoc                    2.7.3                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
pango                     1.42.4               ha030887_1    conda-forge
parso                     0.5.1                      py_0    conda-forge
partd                     1.0.0                      py_0    conda-forge
pcre                      8.43                 he1b5a44_0    conda-forge
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pillow                    6.2.0            py37h6b7be26_0    conda-forge
pip                       19.3.1                   py37_0    conda-forge
pixman                    0.38.0            h516909a_1003    conda-forge
pluggy                    0.13.0                   py37_0    conda-forge
pooch                     0.5.2                    py37_0    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.10                     py_0    conda-forge
psutil                    5.6.3            py37h516909a_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py                        1.8.0                      py_0    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pydocstyle                4.0.1                      py_0    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pyilluminate              0.4.2+0.g1a773c2.dirty           dev_0    <develop>
pylint                    2.4.3                    py37_0    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyqt                      5.9.2            py37hcca6a23_4    conda-forge
pyrsistent                0.15.4           py37h516909a_0    conda-forge
pyserial                  3.4                        py_2    conda-forge
pyside2                   5.9.0a1          py37h4dc837a_0    conda-forge
pysocks                   1.7.1                    py37_0    conda-forge
pytest                    5.2.1                    py37_0    conda-forge
pytest-forked             1.0.2                      py_0    conda-forge
pytest-qt                 3.2.2                      py_0    conda-forge
pytest-xdist              1.30.0                     py_0    conda-forge
python                    3.7.3                h33d41f4_1    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
python-graphviz           0.13                       py_0    conda-forge
python-jsonrpc-server     0.2.0                      py_0    conda-forge
python-language-server    0.29.1                   py37_0    conda-forge
pytoml                    0.1.21                     py_0    conda-forge
pytz                      2019.3                     py_0    conda-forge
pywavelets                1.1.1            py37hc1659b7_0    conda-forge
pyyaml                    5.1.2            py37h516909a_0    conda-forge
pyzmq                     18.1.0           py37h1768529_0    conda-forge
qdma                      0.0.0+242.gea8cc6b           dev_0    <develop>
qt                        5.9.7                h0c104cb_3    conda-forge
qtawesome                 0.6.0                      py_0    conda-forge
qtconsole                 4.5.5                      py_0    conda-forge
qtpy                      1.9.0                      py_0    conda-forge
qtpy-uic                  2.0.4                      py_0    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
requests                  2.22.0                   py37_1    conda-forge
requests_download         0.1.2                      py_1    conda-forge
rope                      0.14.0                     py_0    conda-forge
scikit-image              0.16.1           py37hb3f55d8_0    conda-forge
scipy                     1.3.1            py37h921218d_2    conda-forge
secretstorage             3.1.1                    py37_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.4.0                   py37_0    conda-forge
sip                       4.19.8          py37hf484d3e_1000    conda-forge
six                       1.12.0                py37_1000    conda-forge
snowballstemmer           2.0.0                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
sphinx                    2.2.0                      py_0    conda-forge
sphinx_rtd_theme          0.4.3                      py_0    conda-forge
sphinxcontrib-applehelp   1.0.1                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.1                      py_0    conda-forge
sphinxcontrib-htmlhelp    1.0.2                      py_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.2                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.3                      py_0    conda-forge
spyder                    3.3.6                    py37_1    conda-forge
spyder-kernels            0.5.2                    py37_0    conda-forge
sqlite                    3.30.1               hcee41ef_0    conda-forge
swig                      4.0.1                he1b5a44_0    conda-forge
tblib                     1.4.0                      py_0    conda-forge
teensy_loader_cli         2.1.1                h14c3975_0    local
terminado                 0.8.2                    py37_0    conda-forge
testpath                  0.4.2                   py_1001    conda-forge
tifffile                  0.15.1          py37h3010b51_1001    conda-forge
tk                        8.6.9             hed695b0_1003    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.3            py37h516909a_0    conda-forge
tqdm                      4.36.1                     py_0    conda-forge
traitlets                 4.3.3                    py37_0    conda-forge
tuna                      0.3.2                      py_0    conda-forge
typed-ast                 1.4.0            py37h516909a_0    conda-forge
typing_extensions         3.7.4                    py37_0    conda-forge
urllib3                   1.25.6                   py37_0    conda-forge
vispy                     0.6.1            py37hf5ce8a7_0    conda-forge
wabisabi                  0.2.8                      py_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.33.6                   py37_0    conda-forge
widgetsnbextension        3.5.1                    py37_0    conda-forge
wrapt                     1.11.2           py37h516909a_0    conda-forge
wurlitzer                 1.0.3                    py37_0    conda-forge
x264                      1!152.20180806       h14c3975_0    conda-forge
xarray                    0.14.0                     py_0    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.9                h516909a_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxpm               3.5.12            h516909a_1002    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-libxt                1.1.5             h516909a_1003    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
yaml                      0.1.7             h14c3975_1001    conda-forge
yapf                      0.28.0                     py_0    conda-forge
zeromq                    4.3.2                he1b5a44_2    conda-forge
zict                      1.0.0                      py_0    conda-forge
zipp                      0.6.0                      py_0    conda-forge
zlib                      1.2.11            h516909a_1006    conda-forge
zstd                      1.4.3                h3b9ef0a_0    conda-forge
```
</details>

**Comment by djhoese** [comment]
@hmaarrfk Is this a 6.0 change? Ideas for alternatives?

**Comment by hmaarrfk** [comment]
I asked upstream


https://github.com/ipython/ipython/issues/11920

**Comment by hmaarrfk** [comment]
We should probably use matplotlibs detection, and maybe allow users to change display frontends.

It is really annoying to do a ton of work, only to have to restart the kernel

**Comment by djhoese** [comment]
> We should probably use matplotlibs detection, and maybe allow users to change display frontends.

What does that mean exactly for vispy?

**Comment by hmaarrfk** [comment]
it could mean that maybe we should simply call their function
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/cbook/__init__.py#L39

or vendor it.

**Comment by hmaarrfk** [comment]
Not too sure how you feel about vendoring matplotlib code

image

But it address my spyder concerns, while keeping the old behavior for notebooks

image

I'm really unable to install the extension for vispy in development mode using conda.... so hard. not sure why.

**Comment by djhoese** [comment]
I am ok with this, but would like the license to go in the module docstring after the explanation you added for where the code came from. The way you have it *might* be better in that it won't get auto-included in the API sphinx docs, but maybe that's a bad thing? Normally the vispy modules start with:

```
# -*- coding: utf-8 -*-
# Copyright (c) Vispy Development Team. All Rights Reserved.
# Distributed under the (new) BSD License. See LICENSE.txt for more info.
```

But I'm not sure how copyright works for this. I'm also not sure how the license works for this. Obviously I want to keep their license, but if their license allows us to use their code and it lives in vispy then is it somehow effected by our vispy license at all?

**Comment by hmaarrfk** [comment]
I can put it in the docstrings, but I just hate seeing the wall of text.

Regarding licensing, we could just agree as vuspy developers, to distribute this file under the matplotlib license and keep them as the sole license owners of the file.

That said, we did extract it and isolate it so there could be some kind of creative value that we added. We could ask them to release the license to us, but again that is unclear. 

We could issue it under a combine license I suppose. 

Finally, it is really unclear who "the vispy developers are" legally, I don't think it exists as an entity. So who would they sue???? You and I??? Maybe....

The simplest would be to create a combined repo, that stays up to date with many popular graphical event loops.

**Comment by djhoese** [comment]
I agree on the wall of text, but I think most people look for license information at the top of a file. Let's leave the license separate from vispy's stuff.

The idea for "vispy developers", which I think I took the idea of from another project, is that the copyright is for whoever modified the file. The exact contributors to a file can be determined by using git so "vispy developers" is shorthand for that. Legally ambiguous, yup. Easy to write, yes.

I still think I'd rather have the license moved *near* the top. If not in the module docstring, then at least before the imports and function.

**Comment by hmaarrfk** [comment]
Moving it to the top will happen.

That said, we might want to move it to a combined license file.

When packaging software, it is annoying to have to manually dig through everything. 
We might need to put it in a combined license file, that indicates which files are controlled by which licenses.

Or, maybe we can work with matplotlib to have a new package that helps with this common backend detection?

Are you still trying to have better matplotlib integration?

**Comment by hmaarrfk** [comment]
I guess you don't want to depend on matplotlib?

**Comment by djhoese** [comment]
I would prefer not to depend on matplotlib as a hard dependency. A combined license file may be fine, but for such a simple function I'm not sure its worth it. What if this module was moved to the `vispy/ext` subpackage where all other vendored packages are?

**Comment by djhoese** [comment]
I thought about this some more. Shouldn't/couldn't this be part of jupyter itself?

**Comment by hmaarrfk** [comment]
This is the worst kind of code to maintain.

I'm not sure anybody wants to support this automagic behavior. I don't... But I know I will include in my library. Or call matplotlib's function

**Comment by djhoese** [comment]
@hmaarrfk Heard of any updates on better ways to do this with changes in the last 5 months? Or do you think this is still the best approach?

**Comment by hmaarrfk** [comment]
Thanks for triaging.

I'm still not sure unfortunately.

I totally understand why you have decided to enable webgl by default.

And with Spyder enabling more notebook like features in the near future, I'm not sure where Qt + GL stands.

Since this is all about user experience, i fear that I will undoubtadly break somebody's system doing this

**Comment by djhoese** [comment]
@hmaarrfk I'm going to try to get a 0.7 release out this month and we have this on the 0.7 milestone. Let's make a final decision on this or at least a plan for the future.

The summary of the problem is that spyder uses jupyter/ipython kernels internally so the current implementation in vispy detects this as a "notebook" and tries to use the webgl backend. This fails in spyder. This PR uses a function copied from matplotlib to detect if a GUI event loop has been initialized. If no event loop exists then we must be in a basic notebook (no GUI) environment and should then use the webgl backend.

The cons of this approach is that vispy would now have to maintain code that it does not "own" and should include the matplotlib license in that bit of code. Other than that though, this is a pretty good solution...right? Are there any other downsides?

**Comment by almarklein** [comment]
I expect that code to be very low-maintenance, so sounds like a good solution.

Haas using env vars be considered, like Carlos [suggested](https://github.com/vispy/vispy/issues/1708#issuecomment-531255461)? E.g. `if "SPY_SYMPY_O" in os.environ): ...`

**Comment by djhoese** [comment]
Using environment variables would tell us what GUI is running VisPy, but it doesn't help in cases where a jupyter notebook starts a Qt event loop (to have Qt windows opened from the browser instead of inline/webgl). At least I think I'm right on that.

**Comment by almarklein** [comment]
Ah ... right.

**Comment by djhoese** [comment]
@hmaarrfk You mention in your comments that this approach doesn't work for some situations. Seeing as this code is just adding extra filtering and these situations probably didn't work that well before, maybe that's OK?

@almarklein How do you feel about all this? The matplotlib code being added. The mpl license being at the top or the bottom of the module. The approach taken to try to detect GUI event loops and jupyter kernels. All of it. If we remove the long `"""` comment, I'm OK not adding a test for this specifically as I agree with @hmaarrfk that this is pretty difficult to test in all situations. Especially since what we really want to test depends on these third-party libraries/applications and how they present themselves to jupyter and GUI event loops.

**Comment by almarklein** [comment]
As far as I understand, this seems to be the best solution. Might not be perfect but at least it should be much better than before. Let's remove that long comment and get this merged.

File: vispy/app/application.py [hunk 1]
```diff
@@ -124,11 +125,53 @@ def is_interactive(self):
...
+        10. Select the Qt5 backend.
+        11. Restart the kernel.
+        12. Run the same code as above (without specifying `%gui qt`)
+        13. Verify that it should run in a seperate window
+        14. Open a `jupyter qtconsole`
+        15. Try running the same code (Failing for now).
+        16. Open a `jupyter console`
+        17. Try running the same code (Failing for now).
+        """
         try:
```
Comment (djhoese) [inline vispy/app/application.py:160]:
```suggestion
    def is_notebook(self):
        """Determine if the user is executing in a Jupyter Notebook"""
        try:
```

**Comment by djhoese** [comment]
Comment removed, title updated a bit and WIP prefix removed. I think when tests passed this can be merged.

**Comment by djhoese** [comment]
Thanks @hmaarrfk!

**Comment by hmaarrfk** [comment]
man sorry for being MIA on this one.

i was out of the office for a while this month.

glad to see this in the next release.

**Pull request merged**